File: baekjoon/p2108.py
def p2108(nums):
    nums.sort()
    average = sum(nums) // len(nums)
    if ( sum(nums) / len(nums) )% 1 >= 0.5:
        average += 1
    median = nums[len(nums) // 2]

    count = {}
    for num in nums:
        if num not in count:
            count[num] = 0
        count[num] += 1
    count_arr = []
    for num in count:
        count_arr.append((count[num], num))
    count_arr.sort(reverse=True)
    max_count = count_arr[0][0]
    i = 0
    while i < len(count_arr) and count_arr[i][0] == max_count:
        i += 1
    max_count_nums = []
    for j in range(i):
        max_count_nums.append(count_arr[j][1])
    max_count_nums.sort()
    if len(max_count_nums) > 1:
        mode = max_count_nums[1]
    else:
        mode = max_count_nums[0]

    r = nums[-1] - nums[0]

    return average, median, mode, r


nums = []
for _ in range(int(input())):
    nums.append(int(input()))

for ans in p2108(nums):
    print(ans)

File: baekjoon/p2108.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <unordered_map>
#include <cmath>
using namespace std;
bool sortByFirstDescAndSecondAsc(const pair<int, int> &a, const pair<int, int> &b){
    if (a.first == b.first) return a.second < b.second;
    return a.first > b.first;
}

int main() {
    int N, first, tmp, sum, average, median, mode, range;
    bool allSame = true;
    vector<int> nums;
    vector< pair<int, int> > countArr;
    unordered_map<int, int> counter;
    cin >> N;
    cin >> first;
    nums.push_back(first);
    for (int i = 1; i < N; i++) {
        cin >> tmp;
        if (tmp != first) allSame = false;
        nums.push_back(tmp);
    }

    // all numbers are same
    if (allSame) {
        cout << first << '\n' 
        << first << '\n'
        << first << '\n'
        << 0 << '\n';
        return 0;
    }


    // average
    sort(nums.begin(), nums.end());
    sum = accumulate(nums.begin(), nums.end(), 0);
    average = round((float)sum / N);
    
    // median
    median = nums[N / 2];

    // mode
    for (int i = 0; i < N; i++) {
        if (!counter.count(nums[i])) counter.insert(make_pair(nums[i], 0));
        counter[nums[i]]++;
    }
    for (auto &element : counter) {
        countArr.push_back(make_pair(element.second, element.first));
    }

    sort(countArr.begin(), countArr.end(), sortByFirstDescAndSecondAsc);
    
    if (countArr[0].first == countArr[1].first) { // no same number -> len(countArr) >= 2
        mode = countArr[1].second;
    } else {
        mode = countArr[0].second;
    }

    // range
    range = nums[N-1] - nums[0];



    cout << average << '\n'
    << median << '\n'
    << mode << '\n'
    << range << '\n';

    return 0;
}